### PR TITLE
fix: dedupe digests before DynamoDB BatchGetItem

### LINKED ISSUE #4279: [Bug]: UI home stays in Loading with DynamoDB — GlobalSearch sends duplicate keys to BatchGetItem
### zot version

v2.1.18 (also reproduced on `main`: the affected function is byte-identical)

### Describe the bug

With `cacheDriver: dynamodb` as remote cache/metaDB, the UI home/explore page hangs on **Loading** forever. The search request returns **HTTP 200** with the error inside the GraphQL body:

```json
{"errors":[{"message":"operation error DynamoDB: BatchGetItem, https response error StatusCode: 400,
  api error ValidationException: Provided list of item keys contains duplicates",
  "path":["GlobalSearch"]}],"data":null}
```

Two things make this hard to notice:

1. **Nothing is logged.** There is no error line in the zot log at all (checked 6h of logs at `log.level: info`) — the error only travels in the GraphQL response body. Monitoring by status code or by log line reports a perfectly healthy registry.
2. **Everything else works.** Pulls, `/v2/_catalog`, `ExpandedRepoInfo(repo:)` and `Image(image:)` are all fine — they don't go through the batched path. Only the global search does, which is what the home page loads.

### Root cause

`fetchImageMetaAttributesByDigest` passes the caller's digest list straight to `BatchGetItem`, chunked by `AwsS3BatchLimit`, without deduplicating:

https://github.com/project-zot/zot/blob/v2.1.18/pkg/meta/dynamodb/dynamodb.go#L2138-L2162

DynamoDB rejects any request containing the same key twice. Repeated digests are legitimate input (several tags, or several repos, referring to one manifest), and **the boltdb driver accepts them** — its `FilterImageMeta` just iterates the list and overwrites the map entry. So the same `MetaDB` call succeeds on boltdb and fails on dynamodb, which is presumably why this survives: default and dev setups use boltdb.

This is the same class of bug as #2464, which was fixed in the image-detail path only. #2629 later touched this very function to add the 100-key chunking, but did not dedupe.

There is also an existing test that **codifies the failure as expected behaviour** — it asserts an error is returned for duplicate digests, using the driver quirk as a convenient error path:

https://github.com/project-zot/zot/blob/v2.1.18/pkg/meta/dynamodb/dynamodb_test.go#L238-L246

### To reproduce

1. Configure zot with `cacheDriver: dynamodb` (S3 storage, `remoteCache: true`).
2. Open the UI home page.
3. The page stays in Loading; the `GlobalSearch` request returns 200 with the payload above.

It reproduces directly in the existing test harness against `amazon/dynamodb-local`, which enforces the same constraint as AWS:

```
DYNAMODBMOCK_ENDPOINT=http://localhost:8000 go test ./pkg/meta/dynamodb/ -run TestWrapperErrors
...
Line 248:
Actual: 'operation error DynamoDB: BatchGetItem, ... ValidationException: Provided list of item keys contains duplicates'
--- FAIL: TestWrapperErrors
```

(after changing that assertion from `ShouldNotBeNil` to `ShouldBeNil`, i.e. asserting the behaviour boltdb already has)

### Expected behavior

The UI home page lists repositories, and `FilterImageMeta` accepts repeated digests on dynamodb exactly as it does on boltdb.

### Additional context

Observed on a registry with ~90 repositories where one repo at a time makes the query fail. Worth noting for anyone hitting this: **deleting the offending repository does not help** — the failure simply moves to another repository, since it depends on which digest list the query happens to build.

PR with the fix and the updated test: linked below.

### PR DESCRIPTION
**Pull request by sfernandez-docline**
**What type of PR is this?**

bug

**Which issue does this PR fix**:

Fixes #4279

**What does this PR do / Why do we need it**:

`fetchImageMetaAttributesByDigest` passes the caller's digest list straight to `BatchGetItem`, chunked by `AwsS3BatchLimit`, without deduplicating. DynamoDB rejects any request containing the same key twice:

```
api error ValidationException: Provided list of item keys contains duplicates
```

Repeated digests are legitimate input — several tags, or several repos, can refer to one manifest — and the **boltdb driver accepts them**: its `FilterImageMeta` iterates the list and overwrites the map entry. So today the same `MetaDB` call succeeds on boltdb and fails on dynamodb.

The user-visible effect is that the UI home/explore page hangs on *Loading* forever: `GlobalSearch` returns **HTTP 200** with the DynamoDB error inside the GraphQL body and `data: null`, and **nothing is written to the zot log**, so the registry looks healthy. Pulls, `/v2/_catalog`, `ExpandedRepoInfo` and `Image` are unaffected — they don't use this batched path.

Same class of bug as #2464 (fixed in the image-detail path only); #2629 later touched this function to add the 100-key chunking but did not dedupe.

This PR queries each distinct digest once. The response is expanded back to the caller's original order and repetitions when building `orderedResp`, so callers see no behaviour change — including `FilterImageMeta`'s `len(imageMetaAttributes) != len(digests)` check, which keeps working because the returned slice still has one entry per requested digest. As a side effect the 100-key chunking now operates on distinct digests, so fewer requests are issued.

**If an issue # is not available please add repro steps and logs showing the issue**:

See #4279.

**Testing done on this change**:

The existing `FilterImageMeta with duplicate digests` test asserted that duplicates return an error, which codified the driver quirk as expected behaviour. It now asserts the call succeeds and returns the requested digest, matching boltdb.

Verified against `amazon/dynamodb-local`, which enforces the same duplicate-key constraint as AWS.

Before the change:

```
DYNAMODBMOCK_ENDPOINT=http://localhost:8000 go test ./pkg/meta/dynamodb/ -run TestWrapperErrors
Line 248:
Actual: 'operation error DynamoDB: BatchGetItem, ... ValidationException: Provided list of item keys contains duplicates'
--- FAIL: TestWrapperErrors (2.71s)
```

After:

```
ok  	zotregistry.dev/zot/v2/pkg/meta/dynamodb	15.210s
ok  	zotregistry.dev/zot/v2/pkg/meta	4.780s
```

`go build` and `go vet` clean on the package.

**Automation added to e2e**:

No new case — an existing unit test covers the path; its assertion was inverted to the correct expectation.

**Will this break upgrades or downgrades?**

No. No storage, schema or API change; only the shape of the request sent to DynamoDB.

**Does this PR introduce any user-facing change?**:

```release-note
Fix the UI home/explore page hanging on "Loading" when using the DynamoDB metaDB: image metadata lookups no longer send duplicate keys to BatchGetItem.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/meta/dynamodb/dynamodb.go [hunk 1]
```diff
@@ -2137,18 +2137,35 @@ const AwsS3BatchLimit = 100
...
 		end := start + size
 
 		resp, err := dwr.Client.BatchGetItem(ctx, &dynamodb.BatchGetItemInput{
 			RequestItems: map[string]types.KeysAndAttributes{
 				dwr.ImageMetaTablename: {
-					Keys: getBatchImageKeys(digests[start:end]),
+					Keys: getBatchImageKeys(uniqueDigests[start:end]),
 				},
 			},
 		})
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/meta/dynamodb/dynamodb.go:2171]: This is a preexisting issue, shall we fix in this PR, or another?

**Comment by rchincha** [comment]
@sfernandez-docline pls look at review comments and also fix ci failures

File: pkg/meta/dynamodb/dynamodb.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vrajashkr) [reply pkg/meta/dynamodb/dynamodb.go:2171]: Iff we need special retry and backoff logic, I would prefer a separate PR for this.

File: pkg/meta/dynamodb/dynamodb.go [hunk 2]
```diff
@@ -2137,18 +2137,35 @@ const AwsS3BatchLimit = 100
...
+			continue
+		}
+
+		seenDigests[digest] = struct{}{}
+		uniqueDigests = append(uniqueDigests, digest)
+	}
+
 	// AWS S3 as a limit (=100) on number of keys that can retrieved in one
 	// request, so break it up
 	batchedResp := []map[string]types.AttributeValue{}
```
[the remark this answers is not in the document]
Comment (vrajashkr) [reply pkg/meta/dynamodb/dynamodb.go]: Existing issue, but it would be much appreciated if the comment could be updated.

[the remark this answers is not in the document]
Comment (sfernandez-docline) [reply pkg/meta/dynamodb/dynamodb.go]: Done in the amended commit — the comment now reads:

```go
// BatchGetItem has a limit (=100) on the number of keys that can be retrieved
// in one request, so break it up
```

Attributing the limit to BatchGetItem rather than S3, and the missing word fixed.

File: pkg/meta/dynamodb/dynamodb.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfernandez-docline) [reply pkg/meta/dynamodb/dynamodb.go:2171]: Agreed it is a real gap, and it predates this change: the `UnprocessedKeys` handling is untouched here. Following @vrajashkr's preference I am leaving retry/backoff out of this PR so it stays a minimal fix for the duplicate-key ValidationException, and I am happy to open a separate issue/PR for it.

Worth noting the two are independent: deduplicating the keys does not change how unprocessed keys are handled, it only stops the request from being rejected outright.

**Comment by sfernandez-docline** [comment]
@rchincha both addressed:

**CI**: the failing check was DCO — my sign-off name did not match the commit author (`sfernandez-docline` vs `Sergio Fernandez`). Amended so both are `Sergio Fernandez <s.fernandez@docline.com>`. Everything else was already green.

**Review comments**:
- The 100-key limit comment is corrected to attribute the limit to `BatchGetItem` instead of S3, plus the missing word — thanks @vrajashkr for calling it out.
- `UnprocessedKeys` retry/backoff is left out of this PR, per @vrajashkr's preference, and it is untouched by this change. Happy to open a separate issue for it if useful.

The fix itself is unchanged. For context on how this shows up in production: the metaDB is shared between two replicas over S3, and `GlobalSearch` reliably hit this because one repo in the catalogue was enough to poison the whole query — the UI home page hung on "Loading" while pulls, `/v2/_catalog`, `ExpandedRepoInfo` and `Image` all worked, and nothing was logged (the error only travels in the GraphQL body, with HTTP 200). Deleting the offending repository just moved the failure to another one.

**Comment by andaaron** [comment]
@sfernandez-docline, please rebase and sign the commit https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits.

**Comment by rchincha** [comment]
`git commit --amend --no-edit -S -s`

**Comment by sfernandez-docline** [comment]
@andaaron @rchincha done: rebased onto current `main` (it was 3 commits behind) and the commit is now signed with SSH (ed25519), single sign-off, author and sign-off matching so DCO passes.

One caveat I want to be upfront about: GitHub currently reports the signature as `unknown_key` because the signing key is not registered on my account yet. I am adding it — once it is there this same commit will show as verified, no force-push needed.

No functional change in this round: the rebase was clean and the only content change since your review is the corrected comment above the chunking loop.

**Comment by andaaron** [comment]
@sfernandez-docline please also follow these steps https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account

**Comment by sfernandez-docline** [comment]
The commit now shows as verified — thanks for the pointer.

On the two red checks: I believe both are unrelated to this change.

`Run zot with extensions tests` fails in `TestDedupeRestoreCompleteMarker` (`pkg/storage/local/local_test.go:1617`):

```
Expected '..."message":"no repositories found in storage, finished.","component":"dedupe"...'
  to contain substring 'restore-complete marker written' (but it didn't)!
```

The dedupe generator ran before the test's repositories were visible, so the marker was never written — it reads as timing-dependent. `Collect all test coverage` then fails as a consequence of that job.

Why I do not think it comes from this PR:

- The diff only touches `pkg/meta/dynamodb`; the failing test exercises local storage with boltdb.
- The same job passed on the previous head of this PR, with identical code — the only content change since is the corrected comment above the chunking loop.
- The rebase pulled in 32f8533, 2fe9b95 and fcf89d2, none of which touch `pkg/storage`, the scheduler or `pkg/meta`.
- The test itself predates all of this (added in #4089).

Could you re-run the failed jobs? I do not have permission to re-run them on this repo. If it reproduces I am happy to dig into it, though it would likely belong in its own issue.

**Comment by andaaron** [comment]
@sfernandez-docline thank you.
